src/Components/Web/src/Forms/EditForm.cs

```diff
@@ -119,17 +134,46 @@ protected override void BuildRenderTree(RenderTreeBuilder builder)
         // optimizing for the common case where _editContext never changes.
         builder.OpenRegion(_editContext.GetHashCode());
 
-        builder.OpenElement(0, "form");
-        builder.AddMultipleAttributes(1, AdditionalAttributes);
-        builder.AddAttribute(2, "onsubmit", _handleSubmitDelegate);
-        builder.OpenComponent<CascadingValue<EditContext>>(3);
-        builder.AddComponentParameter(4, "IsFixed", true);
-        builder.AddComponentParameter(5, "Value", _editContext);
-        builder.AddComponentParameter(6, "ChildContent", ChildContent?.Invoke(_editContext));
-        builder.CloseComponent();
-        builder.CloseElement();
+        if (FormHandlerName != null)
+        {
+            builder.OpenComponent<CascadingModelBinder>(0);
+            builder.AddComponentParameter(1, nameof(CascadingModelBinder.Name), FormHandlerName);
+            builder.AddComponentParameter(2, nameof(CascadingModelBinder.ChildContent), (RenderFragment<ModelBindingContext>)RenderWithNamedContext);
+            builder.CloseComponent();
+        }
+        else
+        {
+            RenderFormContents(builder, BindingContext);
+        }
 
         builder.CloseRegion();
+
+        RenderFragment RenderWithNamedContext(ModelBindingContext context)
+        {
+            return builder => RenderFormContents(builder, context);
+        }
+
+        void RenderFormContents(RenderTreeBuilder builder, ModelBindingContext? bindingContext)
+        {
+            builder.OpenElement(0, "form");
+            if (!string.IsNullOrEmpty(bindingContext?.Name))
+            {
+                builder.AddAttribute(1, "name", bindingContext.Name);
+                builder.AddAttribute(2, "action", $"?handler={bindingContext.Name}");
+            }
+            builder.AddMultipleAttributes(3, AdditionalAttributes);
+            builder.AddAttribute(4, "onsubmit", _handleSubmitDelegate);
+            if (bindingContext != null)
+            {
+                builder.SetEventHandlerName(bindingContext.BindingContextId);
```

Comment:
To what extent are you happy with the limitation that SSR event handling only works with `EditForm` and not with a basic `<form>`, even though a basic `<form>` has an `@onsubmit` that looks as if it should also work? This also implies it doesn't work with any custom `MyForm` component that wraps `<form>` without wrapping `EditForm`.

The main alternative I can think of is changing how the named-event tracking works so that:

 * For renderers with some flag enabled (which we only enable for `EndpointHtmlRenderer`), `ComponentState` is updated to track the list of all event names used by a component on its most recent render. There would be no `SetEventHandlerName` API.
 * Then, in `EndpointHtmlRenderer`'s `UpdateDisplayAsync`, it aggregates this to pick out all the `onsubmit` events into a list of component IDs that are currently registering one of those events
 * Then, when trying to dispatch a form post, `EndpointHtmlRenderer` looks at all the components that currently have an `onsubmit` event, and walks through their current rendertree to find the one matching the desired name and its event handler ID

I know the event dispatch phase would involve more processing, but it's hard to see it being problematic because (1) a relatively tiny number of components will register an `onsubmit` event out of the complete set we just rendered, and (2) walking the current rendertree for a component is far less expensive than rendering and diffing it, so this has no impact on the big-O complexity of what we're doing altogether.

This would have the benefit of enabling form-post handling for `<form>` and making it innately work with any `onsubmit` handler that could work during regular interactive rendering.

Is it worth it? I'm open to going either way. I also understand you may not want to make such a big reshuffle before merging this PR. But if you think it's potentially a good improvement for the future, it would be good enough to file an issue and we can cover it in a later preview. I'd be happy to implement this if you want.

Reply:
Filed https://github.com/dotnet/aspnetcore/issues/47804 to track revisting this aspect.